python/python/bystro/covariance/_base_covariance.py

```diff
@@ -0,0 +1,549 @@
+    """
+    if weights is None:
+        weights = np.ones(X.shape[0])
+    avg_score = np.mean(weights * _score_samples(covariance, X))
+    return avg_score
+
+
+def _score_samples(covariance, X):
+    """
+    Return the log likelihood of each sample
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    X : np.array-like,(N,sum(p))
+        The data
+
+    Returns
+    -------
+    scores : float
+        Log likelihood for each sample
+    """
+    p = covariance.shape[0]
+    term1 = -p / 2 * np.log(2 * np.pi)
+
+    _, logdet = la.slogdet(covariance)
+    term2 = -0.5 * logdet
+
+    quad_init = la.solve(covariance, np.transpose(X))
+    difference = X * np.transpose(quad_init)
+    term3 = np.sum(difference, axis=1)
+
+    scores = term1 + term2 - 0.5 * term3
+    return scores
+
+
+#####################################
+#####################################
+##                                 ##
+##  Information-theoretic methods  ##
+##                                 ##
+#####################################
+#####################################
+
+
+def _entropy(covariance):
+    """
+    Computes the entropy of a Gaussian distribution parameterized by
+    covariance.
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    Returns
+    -------
+    entropy : float
+        The differential entropy of the distribution
+    """
+    covariance.shape[1]
+    cov_new = 2 * np.pi * np.e * covariance
+    _, logdet = la.slogdet(cov_new)
+    entropy = 0.5 * logdet
+    return entropy
+
+
+def _entropy_subset(covariance, idxs):
+    """
+    Computes the entropy of a subset of the Gaussian distribution
+    parameterized by covariance.
+
+    Parameters
+    ----------
+    covariance : np.array-like(p,p)
+        The covariance matrix
+
+    idxs: np.array-like,(sum(p),)
+        The observation locations
+
+    Returns
+    -------
+    entropy : float
+        The differential entropy of the distribution
+    """
+    cov1 = covariance[idxs == 1]
+    cov_sub = cov1[:, idxs == 1]
+    entropy = _entropy(cov_sub)
+    return entropy
+
+
+def _mutual_information(covariance, idxs1, idxs2):
+    """
+    This computes the mutual information bewteen the two sets of
+    covariates based on the model.
+
+    Parameters
+    ----------
+    idxs1 : np.array-like,(p,)
+        First group of variables
+
+    idxs2 : np.array-like,(p,)
+        Second group of variables
+
+    Returns
+    -------
+    mutual_information : float
+        The mutual information between the two variables
+    """
+    idxs = idxs1 + idxs2
+    cov_sub = covariance[np.ix_(idxs == 1, idxs == 1)]
+    idxs1_sub = idxs1[idxs == 1]
+    Hy = _entropy(cov_sub[np.ix_(idxs1_sub == 1, idxs1_sub == 1)])
+
+    _, covariance_conditional = _conditional_parameters(cov_sub, 1 - idxs1_sub)
+    Hyx = _entropy(covariance_conditional)
+    mutual_information = Hy - Hyx
```

Comment:
This line ended up being surprising to me because I instinctively read `Hyx` as the joint entropy of X and Y rather than the conditional entropy H(y|x).  Could we rename to `H_y_given_x` or similar?

Reply:
Done.